src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,505 @@
+      on empirical observations but in reality this could be larger for more
+      complex domains. The downside if this is too large is that real anomalies
+      might get ignored and not flagged.
+
+    @param estimationSamples - (int) the number of reasonable anomaly scores
+      required for the initial estimate of the Gaussian. The default of 100
+      records is reasonable - we just need sufficient samples to get a decent
+      estimate for the Gaussian. It's unlikely you will need to tune this since
+      the Gaussian is re-estimated every 10 iterations by default.
+
+    @param historicWindowSize - (int) size of sliding window of historical
+      data points to maintain for periodic reestimation of the Gaussian. Note:
+      the default of 8640 is based on a month's worth of history at 5-minute
+      intervals.
+
+    @param reestimationPeriod - (int) how often we re-estimate the Gaussian
+      distribution. The ideal is to re-estimate every iteration but this is a
+      performance hit. In general the system is not very sensitive to this
+      number as long as it is small relative to the total number of records
+      processed.
+
+  **/
+AnomalyLikelihood::AnomalyLikelihood(UInt learningPeriod, UInt estimationSamples, UInt historicWindowSize, UInt reestimationPeriod, UInt aggregationWindow) :
+    learningPeriod(learningPeriod),
+    reestimationPeriod(reestimationPeriod),
+    averagedAnomaly(aggregationWindow) {
+        iteration = 0;
+        probationaryPeriod = learningPeriod+estimationSamples;
+        assert(historicWindowSize >= estimationSamples); // cerr << "estimationSamples exceeds historicWindowSize";
+        assert(aggregationWindow < reestimationPeriod && reestimationPeriod < historicWindowSize);
+        
+        runningAverageAnomalies.set_capacity(historicWindowSize); 
+        runningLikelihoods.set_capacity(historicWindowSize);
+        runningRawAnomalyScores.set_capacity(historicWindowSize);
+        runningRawValues.set_capacity(historicWindowSize);
+        assert(runningLikelihoods.capacity() == historicWindowSize);
+    }
+
+    
+Real AnomalyLikelihood::anomalyProbability(Real rawValue, Real anomalyScore, int timestamp) { //TODO "rawValue" could be typed to T rawValue; //TODO is rawValue ever used?
+    /**
+    Compute the probability that the current value plus anomaly score represents
+    an anomaly given the historical distribution of anomaly scores. The closer
+    the number is to 1, the higher the chance it is an anomaly.
+
+    @param value - the current metric ("raw") input value, eg. "orange", or
+                   '21.2' (deg. Celsius), ...
+    @param anomalyScore - the current anomaly score
+    @param timestamp - (optional) timestamp of the ocurrence,
+                       default (-1) results in using iteration step.
+    @return the anomalyLikelihood for this record.
+    **/
+    Real likelihood = 0.5f;
+
+    if (timestamp == -1) {
+      timestamp = this->iteration;
```

Comment:
Why is there a preference to a (local) naming convention (localName_) to a self explanatory `this->someVar` ?